File: bindings/dart/lib/src/wasm.dart
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// WASM protocol types for proven-servers.

/// ValType matching the Idris2 ABI tags.
enum ValType {
  i32(0),
  i64(1),
  f32(2),
  f64(3),
  v128(4),
  funcRef(5),
  externRef(6);

  const ValType(this.tag);
  final int tag;

  static ValType? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// ExternKind matching the Idris2 ABI tags.
enum ExternKind {
  funcExtern(0),
  tableExtern(1),
  memExtern(2),
  globalExtern(3);

  const ExternKind(this.tag);
  final int tag;

  static ExternKind? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// Mutability matching the Idris2 ABI tags.
enum Mutability {
  immutable(0),
  mutable(1);

  const Mutability(this.tag);
  final int tag;

  static Mutability? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

File: bindings/javascript/src/wasm.js
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) Jonathan D.A. Jewell <j.d.a.jewell@open.ac.uk>
//
// WASM protocol types for proven-servers.

/** ValType matching the Idris2 ABI tags. */
export const ValType = Object.freeze({
  I32: 0,
  I64: 1,
  F32: 2,
  F64: 3,
  V128: 4,
  FUNC_REF: 5,
  EXTERN_REF: 6,
});

/** ExternKind matching the Idris2 ABI tags. */
export const ExternKind = Object.freeze({
  FUNC_EXTERN: 0,
  TABLE_EXTERN: 1,
  MEM_EXTERN: 2,
  GLOBAL_EXTERN: 3,
});

/** Mutability matching the Idris2 ABI tags. */
export const Mutability = Object.freeze({
  IMMUTABLE: 0,
  MUTABLE: 1,
});
